pkg/link/spoofcheck.go

```diff
@@ -102,20 +106,31 @@ func (sc *SpoofChecker) Setup() error {
 	rulesConfig.AddRule(sc.matchMacRule(macChain.Name))
 	rulesConfig.AddRule(sc.dropRule(macChain.Name))
 
-	if err := sc.configurer.Apply(rulesConfig); err != nil {
+	rulestore, err := sc.configurer.Apply(rulesConfig)
+	if err != nil {
 		return fmt.Errorf("failed to setup spoof-check: %v", err)
 	}
+	sc.rulestore = rulestore
 
 	return nil
 }
 
+func (sc *SpoofChecker) StoredRules() (*nft.Config, error) {
```

Comment:
> This seems to be acting as a cache. Assuming I understand correctly, here are a few observations about this:

Correct, it caches the previously applied ruleset snippet.

> * As it returns from the store or from an actual full read, I would just name the method `Rules`/`Ruleset`. The caller does not know from where the data came from.

Fine with me! I'd go with 'appliedRules' instead, because of the following:

> * A call to this method seems to invalidate the "cache" as a side effect. Is there a reason for this behavior?

Since Teardown uses the stored rules for a delete command, once used the cache is no longer valid. I wanted to make sure it is not reused by accident. Maybe introduce a second method to invalidate the cache?

>   From a regular cache, I would expect invalidation and refresh to occur when the ruleset is updated (e.g. when `Setup` and `Teardown` is called), but not when we fetch it.

OK, so Teardown() should call sc.appliedRules() && sc.dropCache() or something.

>   Also, when reading the (actual non-cached) ruleset, the cache can be updated so it can be used later again.

That's an interesting point. My original StoredRules() is tailored to the use-case, I merely introduced it to simplify Teardown()'s body a bit. Depending on whether a cache is present or not, it will return different results: The cache will contain the flush commands and the three rule add commands captured in Setup(). If there's no cache (for whatever reason), the whole nat prerouting chain is returned instead.

Maybe I should move the rule searching logic into it, too. Then it would return
either the prerouting rule to delete.

> If the intention was different, sorry in advance.

Sorry for the unclear code. I'm entirely unfamiliar with typical coding style in Go.

Reply:
Sounds good to me.

I can share that in some projects, we created cache wrappers to the backed data (in this case, the `sc.configurer`). So the `Read` and `Apply` got wrapped by the cache object and the logic of when to save the data in the cache has been part of its logic. I guess it is like a cache adapter.
But I think the approach here is good enough, it does the same without wrapping the backend data source, but by operating alongside it.